openscenegraph: add @3.4.0, @3.4.1; add '+ffmpeg'

xjrc opened the pull request:
This pull request makes the following changes to the `openscenegraph` package:

- Adds older versions `@3.4.0` and `@3.4.1`.
- Adds the `+ffmpeg` variant to enable conditionally building with OpenSceneGraph's `ffmpeg` plugin.
- Adds the previously missing `jasper` dependency to this package.

I've verified that the following variants of this package build with `gcc@4.9.3` on architecture `linux-rhel7-broadwell` using  `qt@5.12.7+shared+opengl+webkit+dbus` as a `qt` base:

- `openscenegraph@3.4.0+ffmpeg ^ffmpeg@4.4` (verified that this **doesn't** work)
- `openscenegraph@3.4.0{+|~}ffmpeg ^ffmpeg@2.8.15` (verified that this **does** work)
- `openscenegraph@3.4.1{+|~}ffmpeg ^ffmpeg@4.4`
- `openscenegraph@3.6.4{+|~}ffmpeg ^ffmpeg@4.4`

In order to build `openscenegraph@3.4.0` the latest `ffmpeg` changes are required (see #16359).

Also, tagging @glennpj again as he's the most recent contributor to this package!

xjrc committed:
openscenegraph: add @3.4.0/3.4.1, update versioning to use git, improve dependency handling (esp. for ffmpeg, opengl)

xjrc committed:
openscenegraph: add '+ffmpeg' variant, add 'jasper' dependency, clean up cmake config variables

adamjstewart commented on var/spack/repos/builtin/packages/openscenegraph/package.py:
```diff
@@ -12,41 +11,59 @@ class Openscenegraph(CMakePackage):
 
     homepage = "http://www.openscenegraph.org"
-    url      = "https://github.com/openscenegraph/OpenSceneGraph/archive/OpenSceneGraph-3.6.4.tar.gz"
+    git_url  = "https://github.com/openscenegraph/OpenSceneGraph.git"
```
```suggestion
    git  = "https://github.com/openscenegraph/OpenSceneGraph.git"
```
If you do this, you don't need to specify `git=git_url` for every version. Although I'm not sure how I feel about switching from `url` to `git`. Checksums are more reliable than tags which can change.

xjrc replied:
I wasn't aware git packages could have the URL specified that way; that makes things much simpler! As for `url` versus `git` specification, I rather like using `git` tags for the readability and ease of use benefits, but you do have a good point about the reliability benefits of checksums. I consider reliability to be of much higher importance for something like a Spack package, so I'll revert this change and add checksums for the new versions.

xjrc committed:
openscenegraph: reverted fetches to use checksums to improve reliability

glennpj commented:
This might require an explicit `dependends_on('libpng')`. My most recent builds links against spack built libpng from the dependency graph but cmake detected the system libpng. That might be worth checking out.

xjrc commented:
I saw this as well when I was testing builds. It looks to me like `libpng` can be included as a dependency at least transitively (and it seems you've experienced that as well), but it's best to make it explicit just in case there are version/variant combinations that prevent this. I'll add an update and re-test presently.

xjrc committed:
openscenegraph: add a few more explicit dependencies (e.g. png, tiff, glib).

xjrc commented:
I've updated the dependencies based on what I saw in the build logs and re-tested the builds for the `openscenegraph@3.4.1~ffmpeg` and `openscenegraph@3.6.4~ffmpeg` variants. The "required dependencies" I added are those reported during the configuration process of `@3.4.1`, but a more comprehensive list that takes into account required versus optional dependencies could be compiled by using the `@3.6.X` logs. These logs indicate that libraries like `libpng` and `libtiff` are actually optional (along with most other libraries; all but `qt` it seems), but I'm not certain if this holds for all versions of the library.

I'm just going to leave all of these new libraries as requirements to be on the safe side for now. If someone needs variants for these other "optional" libraries, they could be added using the new  `+ffmpeg` variant as a template.
